Add/remove workflow step [PREMIUM-194]

JanJakes opened the pull request:
Sorry for another huge PR. This got much bigger than I thought. I tried hard to make the commits nice so going commit-by-commit should make sense. Premium PR included as well.

**Testing:** With Premium activated and a valid Premium key, it should be possible to add/remove steps and triggers within a workflow and save the workflow.

[PREMIUM-194]
[MAILPOET-4585]

[PREMIUM-194]: https://mailpoet.atlassian.net/browse/PREMIUM-194?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MAILPOET-4585]: https://mailpoet.atlassian.net/browse/MAILPOET-4585?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

JanJakes committed:
Set step separator height as per specs

[PREMIUM-194]

JanJakes committed:
Add workflow-ending node

[PREMIUM-194]

JanJakes committed:
Make testing button create an empty workflow

[PREMIUM-194]

JanJakes committed:
Implement add trigger placeholder & button

[PREMIUM-194]
[MAILPOET-4585]

JanJakes committed:
Add type (triggers vs. other steps) to inserter popover state

[PREMIUM-194]
[MAILPOET-4585]

JanJakes committed:
Add someone subscribed trigger type

[PREMIUM-194]
[MAILPOET-4585]

JanJakes committed:
Add WP user registered trigger type

[PREMIUM-194]
[MAILPOET-4585]

JanJakes committed:
Implement displaying triggers in step inserter list

[PREMIUM-194]
[MAILPOET-4585]

JanJakes committed:
Remove mocked data, fallback to unknown step type rendering

[PREMIUM-194]

JanJakes committed:
Add cursor: pointer to add step button

[PREMIUM-194]

JanJakes committed:
Fix step list item props type

[PREMIUM-194]

JanJakes committed:
Add onInsert prop to inserter component

[PREMIUM-194]

JanJakes committed:
Show premium upsell modal when trying to add a step

[PREMIUM-194]

JanJakes committed:
Fix Gutenberg block inserter overlaying modal dialog

[PREMIUM-194]

JanJakes committed:
Add step dropdown menu with delete button

[PREMIUM-194]

JanJakes committed:
Disable inserter sidebar until we implement drag & drop

[PREMIUM-194]

JanJakes committed:
Show premium upsell modal when trying to delete a step

[PREMIUM-194]

JanJakes committed:
Remove side-effect from store module, control store initialization

This is important so we're able to export it and use it in the premium plugin.

[PREMIUM-194]

JanJakes committed:
Add hook for editor store extensibility

This will be used in the premium plugin.

[PREMIUM-194]

JanJakes committed:
Add some automation symbols to exported modules

This is needed for the premium plugin.

[PREMIUM-194]

JanJakes committed:
Add step ID to data attributes for workflow steps and separators

[PREMIUM-194]

JanJakes committed:
Make step rendering configurable with hook

[PREMIUM-194]

JanJakes committed:
Make step separator rendering configurable with hook

[PREMIUM-194]

JanJakes committed:
Export StoreKey type for extensibility

[PREMIUM-194]

JanJakes committed:
Add empty trigger type to automations core

[PREMIUM-194]

JanJakes committed:
Make workflow editing logic extensible

[PREMIUM-194]

JanJakes committed:
Add tracking parameters

[PREMIUM-194]

JanJakes committed:
Hotfix inconsistent @wordpress/icon sizings

[PREMIUM-194]

JanJakes committed:
Add automation registry context to editor state

[PREMIUM-194]

JanJakes committed:
Fix workflow run getter and from array construction

[PREMIUM-194]

websupporter commented on mailpoet/lib/Automation/Engine/API/API.php:
```diff
@@ -58,24 +58,29 @@ private function registerRoute(string $route, string $endpointClass, string $met
+        },
+        'args' => $schema,
+      ],
+      true // allow overriding
```
It seems, allowing to overwrite in this way removes the free plugin `GET workflows` in favor of the the premium `POST workflows` route.

At least in my tests, setting the allow overwrite value to `false` recovers the `GET` endpoint

![image](https://user-images.githubusercontent.com/6458412/188061247-b82d65ad-ae5c-49bb-99ea-a8452cf456a6.png)

JanJakes replied:
Yep, good catch! I need to overwrite the PUT route but not the others, damn. I guess I can use some filter to replace it instead. Checking...

JanJakes replied:
Ahh, the fix is actually pretty simple! I'll drop this commit and make the premium route registrations run before the free ones!

websupporter replied:
Thanks! Seems to do the trick 
![image](https://user-images.githubusercontent.com/6458412/188204855-96476b79-1ade-4b2b-b5d5-3cced4cb1bd7.png)

websupporter commented:
General question:
> it should be possible to add/remove [...] triggers

How can I do that? I miss somehow the obvious, but a completely empty workflow seems not to provide me a possibility and I do not see one in a workflow populated with steps
![image](https://user-images.githubusercontent.com/6458412/188064075-5a7db24e-2002-43b1-bab6-581f64a3686a.png)
![image](https://user-images.githubusercontent.com/6458412/188064115-b7b8df2d-372f-4e05-99f3-2dd281308560.png)

websupporter commented:
I wonder, now that there is a empty trigger necessary as a minimum for a workflow; in theory you can send `step:[]` to the API and a workflow is created without an empty trigger and basically useless. I think we should either throw an error from the API for such a case or fill an empty trigger as a default in case of `[]`

![image](https://user-images.githubusercontent.com/6458412/188069216-be6ce3f8-b317-4da6-8ce9-a9f2afd0a1de.png)

websupporter commented on mailpoet/assets/js/src/automation/editor/components/inserter-popover/index.tsx:
```diff
@@ -1,28 +1,66 @@
+import { Item } from '../inserter/item';
+import { storeName } from '../../store';
 
 export function InserterPopover(): JSX.Element | null {
```
`JSX.Element | null`

out of curiosity: Isn't `null` a valid `JSX.Element`?

JanJakes replied:
It's [a bit compicated](https://stackoverflow.com/a/58123882/1236795), but no, `null` is not included in `JSX.Element`, it's only included in `React.ReactNode` but JSX element [seems to be the best consensus](https://github.com/typescript-cheatsheets/react/issues/129) for typing component return types.

JanJakes commented:
> I wonder, now that there is a empty trigger necessary as a minimum for a workflow; in theory you can send step:[] to the API and a workflow is created without an empty trigger and basically useless. I think we should either throw an error from the API for such a case or fill an empty trigger as a default in case of []

Good point! There are more things we don't validate yet, and I want to prepare a ticket for workflow validation, so I would like to include that. I'm not yet fully sure if we'll really need the empty triggers when we improve the workflow format to support multiple next steps (conditions, A/B test, etc.) but I want to look at it next week so I'll keep this in mind.

JanJakes committed:
Set automation newsletters always as active

Their scheduling and validity will handled by the workflow engine so this should be OK.
We can also consider syncing the state based on the workflow state, somehow.

[PREMIUM-194]

JanJakes committed:
Implement sending for automation emails

[PREMIUM-194]

websupporter reviewed (dismissed):
Thanks for the PRs. Really nice work. LGTM
